corehq/apps/case_search/xpath_functions/ancestor_functions.py

```diff
@@ -92,22 +91,73 @@ def _is_ancestor_path_expression(node):
     return isinstance(node.left, Step)
 
 
-def _parent_property_lookup(context, node):
-    """given a node of the form `parent/foo = 'thing'`, return all case_ids where `foo = thing`
+def _child_case_lookup(context, case_ids, identifier):
+    """returns a list of all case_ids who have parents `case_id` with the relationship `identifier`
     """
-    es_filter = property_comparison_query(context, node.left.right, node.op, node.right, node)
-    es_query = CaseSearchES().domain(context.domain).filter(es_filter)
-    if es_query.count() > MAX_RELATED_CASES:
-        new_query = '{} {} "{}"'.format(serialize(node.left.right), node.op, node.right)
-        raise TooManyRelatedCasesError(
-            gettext("The related case lookup you are trying to perform would return too many cases"),
-            new_query
-        )
+    return CaseSearchES().domain(context.domain).get_child_cases(case_ids, identifier).scroll_ids()
 
-    return es_query.scroll_ids()
 
+def ancestor_exists(node, context):
+    """
+    Supports the following syntax:
+    - ancestor_exists({ancestor_path_expression}, {ancestor_filter})
+    i.e. ancestor-exists(host/parent,city='SF')
 
-def _child_case_lookup(context, case_ids, identifier):
-    """returns a list of all case_ids who have parents `case_id` with the relationship `identifier`
+    Since ES has no way of performing joins, we filter down in stages:
+    1. Find the ids of all cases where the condition is met
+    2. Walk down the case hierarchy, finding all related cases with the right identifier to the ids
+    found in (1).
+    3. Return the lowest of these ids as a related case query filter
     """
-    return CaseSearchES().domain(context.domain).get_child_cases(case_ids, identifier).scroll_ids()
+    confirm_args_count(node, 2)
+    ancestor_path_node, ancestor_case_filter_node = node.args
+    return process_ancestor_exists(ancestor_path_node, ancestor_case_filter_node, context)
+
+
+def process_ancestor_exists(ancestor_path_node, ancestor_case_filter_node, context):
+    _validate_ancestor_exists_filter(ancestor_case_filter_node)
+    base_case_ids = _get_case_ids_from_ast_filter(context, ancestor_case_filter_node)
+
+    return walk_ancestor_hierarchy(context, ancestor_path_node, base_case_ids)
+
+
+def _validate_ancestor_exists_filter(node):
+    """
+    ancestor_exists(status="active" and subcase-exists(...)) raises error
+    """
+
+    incompatible_xpath_query_functions = {'subcase-exists', 'subcase-count'}
+
+    if isinstance(node, FunctionCall) and node.name in incompatible_xpath_query_functions:
+        raise CaseFilterError(gettext(f"{node.name} is not supported with ancestor-exists"), node.name)
+
+    if hasattr(node, 'op') and node.op in OPERATOR_MAPPING:
+        # This node is a leaf
+        return OPERATOR_MAPPING[node.op](_validate_ancestor_exists_filter(node.left),
+                                        _validate_ancestor_exists_filter(node.right))
```

Comment:
Other than recursively call the validation, what is the need for the operator mapping here? It doesn't appear that the return value is used at all.

Reply:
you're right, there is no need - [0368c6a](https://github.com/dimagi/commcare-hq/pull/33014/commits/0368c6a215ccd3c891e30c2d609997358c17899c)